pkg/common/cns-lib/volume/manager.go

```diff
@@ -184,161 +192,235 @@ func (m *defaultManager) ResetManager(ctx context.Context, vcenter *cnsvsphere.V
 	log.Infof("Done resetting volume.defaultManager")
 }
 
-// CreateVolume creates a new volume given its spec.
-func (m *defaultManager) CreateVolume(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (*CnsVolumeInfo, error) {
-	internalCreateVolume := func() (*CnsVolumeInfo, error) {
-		log := logger.GetLogger(ctx)
-		err := validateManager(ctx, m)
-		if err != nil {
-			return nil, err
+// createVolumeWithIdempotency leverages the VolumeOperationRequest interface to persist CNS task information. It uses
+// this persisted information to handle idempotency of CreateVolume callbacks to CNS for the same volume.
+func (m *defaultManager) createVolumeWithIdempotency(ctx context.Context, spec *cnstypes.CnsVolumeCreateSpec) (
```

Comment:
With gocsi being removed, would we get multiple invocation of createVolumeWithIdempotency method and we end up creating orphan volumes? Is there a need for serialization per volume request?

Reply:
Is there a scenario where we get concurrent requests to create the same volume?

Reply:
We do not have control on the side cars. What if the side car invokes multiple create volume requests? The CSI API doc say that it could and it expects the CSI driver to handle idempotency.

Reply:
Makes sense. There's a potential window where the first CreateVolume call has not created the `CnsVolumeOperationRequest` object but the second call attempts to `Get()` it (and fails).

Reply:
Since gocsi is currently enabled by default, i'll take this in a follow up change. We'll need to implement serialization for other callbacks too